feat(search): add groups to cmd-k search

thmsobrmlr opened the pull request:
## Problem

People want to search groups.

## Changes

This PR adds groups to the CMD+K search. As in https://github.com/PostHog/posthog/pull/18952, follow-ups are:
- Displaying a better preview
- Adapting the tabs to fit more of them

## How did you test this code?

Clicking around

thmsobrmlr committed:
add groups to api

thmsobrmlr committed:
feat(search): add groups to cmd-k search

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

thmsobrmlr committed:
fix types

daibhin commented on frontend/__snapshots__/components-command-bar--search--dark.png:
I assume these screenshot updates are expected and not just flipping because of a loading state?

daibhin replied:
I was referring to the number counts present in one and not the other. Based on your demo yesterday I suspect it's ok

daibhin commented on frontend/src/lib/components/CommandBar/SearchResult.tsx:
```diff
@@ -69,7 +71,13 @@ export const SearchResult = ({ result, resultIndex, focused, keyboardFocused }:
+                    {result.type !== 'group'
+                        ? tabToName[result.type]
+                        : `${capitalizeFirstLetter(
+                              aggregationLabel((result.extra_fields as Group).group_type_index).plural
```
Is there a reason we need to cast here given we have the `Group` specified in https://github.com/PostHog/posthog/pull/18965/files#diff-ed52228227f4ce03c0bf7b8aa5fcdf749a4fea0b96b4f9b6467b04f3e85a6515R22

thmsobrmlr replied:
Good catch! This was actually a "bug" in the `SearchableEntity` preventing the discriminating union to resolve accordingly. Fix coming in.

daibhin commented on frontend/src/lib/components/CommandBar/constants.ts:
```diff
@@ -14,9 +14,16 @@ export enum Tab {
+    Group1 = 'group_1',
+    Group2 = 'group_2',
+    Group3 = 'group_3',
+    Group4 = 'group_4',
```
why five groups?

daibhin replied:
This seems like the biggest point of confusion / contention here. Is there any means searching for all groups at once and categorising later? We seem add a lot of complexity to the `searchBarLogic` to support this

thmsobrmlr replied:
The search needs to go through 5 different endpoints (one for each possible group). This is also how it works currently. I thought about using an abstraction, but think the verbose is more easily understandable for the future traveller. Note: I'm thinking about adding a query cache in a follow-up, then this might be changed as well.

thmsobrmlr replied:
> why five groups?

(Ignore if you already know this, adding context just in case). PostHog supports up to 5 groups for each project. In the local demo project these are "Organization", "Account", "Instance", "Customer" and "Project" for example. Other groups will be useful for different customers and so we just have "Group 0,1,2,3,4" in our data model. This is the `group_type_index`.
The endpoint we (re-)use here has a mandatory `group_type_index`, thus 5 calls. Also these refer to different entities with different tabs, so a reducer per group is useful anyway.

daibhin replied:
Gotcha! Thanks for the context. Tbh it sounds like a bit of a legacy mess that should just have been a "has-many" relationship 😅 What if I want six groups?

Your approach totally makes sense then

daibhin commented on frontend/src/lib/components/CommandBar/searchBarLogic.ts:
```diff
@@ -116,35 +206,135 @@ export const searchBarLogic = kea<searchBarLogicType>([
+            ): Record<Tab, string | null> => {
                 const counts = {}
-                const personsResults = personsResponse?.results
+                // const personsResults = personsResponse?.results
```
Remove

thmsobrmlr replied:
👌

daibhin commented on frontend/src/lib/components/CommandBar/searchBarLogic.ts:
```diff
@@ -116,35 +206,135 @@ export const searchBarLogic = kea<searchBarLogicType>([
+                    Tab.Action,
+                    Tab.Person,
+                    Tab.Cohort,
+                    ...Array.from(groupTypes.values()).map(
+                        ({ group_type_index }) => `group_${group_type_index}` as Tab
+                    ),
```
Could possibly be a selector of its own

thmsobrmlr replied:
I'll likely group the tabs in a follow up, leaving this refactor for then.

thmsobrmlr committed:
fix searchable entity type

thmsobrmlr committed:
remove leftover comment

thmsobrmlr commented:
Thanks for the review @daibhin ! I addressed your comments in code & replies. Let me know if you think I should still do something about the verbose groups fetching now.

thmsobrmlr committed:
fix tabbing behaviour

thmsobrmlr committed:
Merge branch 'master' into groups-search

# Conflicts:
#	frontend/__snapshots__/scenes-app-sidepanels--side-panel-settings.png
#	frontend/src/lib/components/CommandBar/searchBarLogic.ts

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots
